h/static/scripts/group-forms/components/CreateGroupForm.tsx

```diff
@@ -42,36 +47,80 @@ function Label({
   );
 }
 
-export default function CreateGroupForm() {
-  const nameId = useId();
-  const descriptionId = useId();
+function NameField() {
+  const id = useId();
+  const [characterCount, setCharacterCount] = useState(0);
+
+  function handleInput(e: any) {
+    setCharacterCount(e.target.value.length);
+  }
+
+  const limit = 25;
+  const error = characterCount > limit;
+
+  return (
+    <div className="mb-4">
+      <Label htmlFor={id} text="Name" required />
+      <Input
+        id={id}
+        onInput={handleInput}
+        autofocus
+        autocomplete="off"
+        required
+        data-testid="name"
+        error={error ? `Must be ${limit} characters or less.` : undefined}
+      />
+      <CharacterCounter
+        value={characterCount}
+        limit={25}
+        testid="charcounter-name"
+        error={error}
+      />
+    </div>
+  );
+}
+
+function DescriptionField() {
```

Comment:
`<DescriptionField />` is pretty much the same thing as `<NameField />`, except instead of an `<Input />` it contains a `<Textarea />`, and the `<Textarea />` has different props than `<NameField />`'s `<Input />`. I've a feeling there's a way to factor this to avoid the duplication, but I can't quite figure out how and I'm not sure if it'll actually be better once done (duplication versus complexity tradeoff). If there's a good way to do this, perhaps I can send a follow-up refactoring PR for it.

Reply:
You can write JSX that selects the component dynamically, by making the symbol after the opening `<` a variable rather than a reference to a fixed function. Under the hood an expression like `<Foo prop={bar}/>` just becomes a function call (`jsx(Foo, { prop: bar })`), so any symbol in scope will work. The only constraint is that it has to start with an upper case letter.

A quick sketch showing a possible generalization:

```tsx
function TextField({
  autofocus = false,
  label,
  limit = 500,
  name,
  required = false,
  type = 'input',
}: {
  autofocus?: boolean;
  label: string;
  limit: number;
  name: string;
  required?: boolean;
  type?: 'input' | 'textarea';
}) {
  const id = useId();
  const [value, setValue] = useState('');

  const error = value.length > limit;

  const InputComponent = type === 'input' ? Input : Textarea;

  return (
    <div className="mb-4">
      <Label htmlFor={id} text={label} required={required} />
      <InputComponent
        name={name}
        id={id}
        onInput={e => setValue((e.target as HTMLInputElement).value)}
        autofocus={autofocus}
        autocomplete="off"
        required={required}
        data-testid="name"
        error={error ? `Must be ${limit} characters or less.` : undefined}
        value={value}
      />
      <CharacterCounter
        value={value.length}
        limit={limit}
        testid="charcounter-name"
        error={error}
      />
    </div>
  );
}
```

Used like so:

```tsx
<TextField name="name" label="Name" limit={25} autofocus required />
<TextField name="description" label="Description" limit={250} type="textarea" />
```

Alternatively you could have two components that are generic over the name and label of the field, but keep separate components for a text line vs a text field to begin with.

Reply:
That worked, thanks: https://github.com/hypothesis/h/pull/8804/commits/456be56753f4898a14d22a13fcf859d30128c552